### PR TITLE
Implement casts between enums and json

### PR DESCRIPTION
**Pull request by msullivan**
Part of #251.

### CODE REVIEW

File: edb/edgeql/compiler/casts.py [hunk 1]
```diff
@@ -127,6 +127,15 @@ def compile_cast(
                 subctx.implicit_id_in_shapes = False
                 subctx.implicit_tid_in_shapes = False
                 viewgen.compile_view_shapes(ir_set, ctx=subctx)
+        elif (orig_stype.issubclass(ctx.env.schema, json_t)
+              and new_stype.is_enum(ctx.env.schema)):
+            # Casts from json to enums need some special handling
+            # here, where we have access to the enum type. Just turn
+            # it into json->str and str->enum.
+            str_typ = ctx.env.get_track_schema_object('std::str')
```
Comment (elprans) [inline edb/edgeql/compiler/casts.py:135]: Use `get_track_schema_type` and then you won't need the `assert`

File: edb/lib/std/25-enumoperators.edgeql [hunk 2]
```diff
@@ -105,3 +105,9 @@ CREATE CAST FROM std::anyenum TO std::str {
     SET volatility := 'IMMUTABLE';
     USING SQL CAST;
 };
+
+# The reverse cast from json to enum is handled in compile_cast
```
Comment (elprans) [inline edb/lib/std/25-enumoperators.edgeql:109]: I think we still want an explicit declaration of the reverse cast. Use `USING SQL EXPRESSION` to mark it as "compiler-handled".

File: tests/test_edgeql_enums.py [hunk 3]
```diff
@@ -188,3 +188,27 @@ async def test_edgeql_enums_assignment_04(self):
...
+    async def test_edgeql_enums_json_cast_01(self):
+        self.assertEqual(
+            await self.con.query(
+                "SELECT <json><test::color_enum_t>'RED'"
+            ),
+            ['"RED"'])
+
+        await self.assert_query_result(
+            "SELECT <test::color_enum_t><json>'RED'",
+            ['RED'])
```
Comment (1st1) [inline tests/test_edgeql_enums.py:201]: `self.assert_query_result` checks both JSON and binary compile modes automatically. SO I'd also add a test without an explicit `<json>` cast, testing that JSON mode handles enums just fine too:

```
        await self.assert_query_result(
            "SELECT <test::color_enum_t>'RED'",
            ['RED'])
```

Comment (msullivan) [reply tests/test_edgeql_enums.py:201]: I can add that test, but it looks like it already passes even without my changes

File: edb/edgeql/compiler/casts.py [hunk 4]
```diff
@@ -128,12 +128,11 @@ def compile_cast(
                 subctx.implicit_tid_in_shapes = False
                 viewgen.compile_view_shapes(ir_set, ctx=subctx)
         elif (orig_stype.issubclass(ctx.env.schema, json_t)
-              and new_stype.is_enum(ctx.env.schema)):
+              and new_stype.is_enum(ctx.env.schema) and False):
```
Comment (elprans) [inline edb/edgeql/compiler/casts.py:131]:
```suggestion
              and new_stype.is_enum(ctx.env.schema)):
```

Comment (msullivan) [reply edb/edgeql/compiler/casts.py:131]: oops! debug code slipped in the last update

Comment (1st1) [reply edb/edgeql/compiler/casts.py]: Was it caught by tests?

Comment (elprans) [reply edb/edgeql/compiler/casts.py]: Yes, tests failed.

**Pull request merged**